src/api.ts

```diff
@@ -1,75 +1,72 @@
 import type { Json } from '@metamask/utils';
 import { JsonStruct } from '@metamask/utils';
 import type { Infer } from 'superstruct';
-import { array, enums, literal, record, string, union } from 'superstruct';
-
+import {
+  enums,
+  array,
+  define,
+  validate,
+  literal,
+  record,
+  string,
+  union,
+  mask,
+} from 'superstruct';
+
+import type { StaticAssertAbstractAccount } from './base-types';
+import type { EthEoaAccount, EthErc4337Account } from './eth';
+import {
+  EthEoaAccountStruct,
+  EthErc4337AccountStruct,
+  EthAccountType,
+} from './eth';
 import { exactOptional, object } from './superstruct';
 import { UuidStruct } from './utils';
 
 /**
- * Supported Ethereum methods.
+ * Type of supported accounts.
  */
-export enum EthMethod {
-  // General signing methods
-  PersonalSign = 'personal_sign',
-  Sign = 'eth_sign',
-  SignTransaction = 'eth_signTransaction',
-  SignTypedDataV1 = 'eth_signTypedData_v1',
-  SignTypedDataV3 = 'eth_signTypedData_v3',
-  SignTypedDataV4 = 'eth_signTypedData_v4',
-  // ERC-4337 methods
-  PrepareUserOperation = 'eth_prepareUserOperation',
-  PatchUserOperation = 'eth_patchUserOperation',
-  SignUserOperation = 'eth_signUserOperation',
-}
+export type KeyringAccounts = StaticAssertAbstractAccount<
+  EthEoaAccount | EthErc4337Account
+>;
 
 /**
- * Supported Ethereum account types.
+ * Mapping between account types and their matching `superstruct` schema.
  */
-export enum EthAccountType {
-  Eoa = 'eip155:eoa',
-  Erc4337 = 'eip155:erc4337',
-}
-
-export const KeyringAccountStruct = object({
-  /**
-   * Account ID (UUIDv4).
-   */
-  id: UuidStruct,
-
-  /**
-   * Account address or next receive address (UTXO).
-   */
-  address: string(),
-
-  /**
-   * Keyring-dependent account options.
-   */
-  options: record(string(), JsonStruct),
-
-  /**
-   * Account supported methods.
-   */
-  methods: array(
-    enums([
-      `${EthMethod.PersonalSign}`,
-      `${EthMethod.Sign}`,
-      `${EthMethod.SignTransaction}`,
-      `${EthMethod.SignTypedDataV1}`,
-      `${EthMethod.SignTypedDataV3}`,
-      `${EthMethod.SignTypedDataV4}`,
-      `${EthMethod.PrepareUserOperation}`,
-      `${EthMethod.PatchUserOperation}`,
-      `${EthMethod.SignUserOperation}`,
-    ]),
-  ),
+export const KeyringAccountStructs: Record<string, any> = {
+  [`${EthAccountType.Eoa}`]: EthEoaAccountStruct,
+  [`${EthAccountType.Erc4337}`]: EthErc4337AccountStruct,
+};
 
-  /**
-   * Account type.
-   */
-  type: enums([`${EthAccountType.Eoa}`, `${EthAccountType.Erc4337}`]),
+/**
+ * Base type with account's type for any account as a `superstruct.object`.
+ */
+export const KeyringAccountTypedStruct = object({
```

Comment:
I think the name `KeyringAccountTypedStruct` is a bit confusing, it looks like it's a `KeyringAccount`, but typed.

Reply:
Yup... I cannot find a better name... Maybe `BaseKeyingAccountStruct`? But I've added a `BaseAccount` type already which differs a bit, so IDK... WDYT?

Reply:
I went for `BaseKeyringAccountStruct`! Lemme know what you think :)